src/components/tabs/Tabs.tsx

```diff
@@ -22,71 +22,85 @@ function isPanelSlotEntry(entry: [string, ComponentChildren]): entry is [PanelSl
 }
 
 function getBaseKeyFromTab(slot: TabSlot) {
-	return slot.replace(new RegExp(`^${tabSlotKey}`), '')
+	return slot.replace(new RegExp(`^${tabSlotKey}`), '');
 }
 
 function getBaseKeyFromPanel(slot: PanelSlot) {
-	return slot.replace(new RegExp(`^${panelSlotKey}`), '')
+	return slot.replace(new RegExp(`^${panelSlotKey}`), '');
 }
 
 type Props = {
 	[key: TabSlot | PanelSlot]: ComponentChildren;
 	sharedStore?: string;
-}
+};
 
 export default function Tabs({ sharedStore, ...slots }: Props) {
-	const tabId = genTabId()
-	const tabs = Object.entries(slots).filter(isTabSlotEntry)
-	const panels = Object.entries(slots).filter(isPanelSlotEntry)
+	const tabId = genTabId();
+	const tabs = Object.entries(slots).filter(isTabSlotEntry);
+	const panels = Object.entries(slots).filter(isPanelSlotEntry);
 	/** Used to focus next and previous tab on arrow key press */
-	const tabButtonRefs = useRef<Record<TabSlot, HTMLButtonElement | null>>({})
-	const scrollToTabRef = useRef<HTMLButtonElement | null>(null)
+	const tabButtonRefs = useRef<Record<TabSlot, HTMLButtonElement | null>>({});
+	const scrollToTabRef = useRef<HTMLButtonElement | null>(null);
+	const tabButtonContainerRef = useRef<HTMLDivElement | null>(null);
+	const activeTabIndicatorRef = useRef<HTMLSpanElement | null>(null);
 
-	const firstPanelKey = panels[0]?.[0] ?? ''
-	const [curr, setCurr] = useTabState(getBaseKeyFromPanel(firstPanelKey), sharedStore)
+	const firstPanelKey = panels[0]?.[0] ?? '';
+	const [curr, setCurr] = useTabState(getBaseKeyFromPanel(firstPanelKey), sharedStore);
 
 	function updateCurr(tabSlot: TabSlot, el: HTMLButtonElement | null) {
 		if (sharedStore) {
 			// Prevents scroll jumping due to layout shift
 			// from other tab views with the same store
-			scrollToTabRef.current = el
+			scrollToTabRef.current = el;
 		}
-		setCurr(getBaseKeyFromTab(tabSlot))
+		setCurr(getBaseKeyFromTab(tabSlot));
 	}
 
 	useEffect(() => {
 		if (scrollToTabRef.current) {
-			scrollToTabRef.current.scrollIntoView({ behavior: 'smooth' })
-			scrollToTabRef.current = null
+			scrollToTabRef.current.scrollIntoView({ behavior: 'smooth' });
+			scrollToTabRef.current = null;
+		}
+	}, [curr]);
+
+	useLayoutEffect(() => {
+		// Fancy indicator animation
+		const activeTab = tabButtonRefs?.current[`tab.${curr}`];
+		if (activeTabIndicatorRef.current && tabButtonContainerRef.current && activeTab) {
+			console.log('A');
+			const tabBoundingRect = activeTab.getBoundingClientRect();
+			const containerBoundingRect = tabButtonContainerRef.current.getBoundingClientRect();
+			activeTabIndicatorRef.current.style.width = tabBoundingRect.width + 'px';
+			activeTabIndicatorRef.current.style.left = tabBoundingRect.left - containerBoundingRect.left + 'px';
```

Comment:
I was worried that animating `left` and `width` had the potential to cause performance issues. This is because changing both of these properties triggers changes to the element geometry and therefore layout calculations. ([CSS Triggers](https://csstriggers.com/) is a useful page for checking this kind of thing.)

After some performance testing, I could see a fair number of frames that are slightly over their time budget during the animation:

<img width="330" alt="image" src="https://user-images.githubusercontent.com/357379/185767365-a2124e61-8679-47ea-a16b-6d5862acf616.png">

It’s not hugely slow (0.1ms over budget), but could be worse on lower powered devices or if something else were happening simultaneously maybe.

We can fix this by switching to using a `transform` to modify the indicator’s position instead:

```suggestion
			if (!activeTabIndicatorRef.current.style.width) activeTabIndicatorRef.current.style.width = '1px';
			activeTabIndicatorRef.current.style.transform = `translateX(${tabBoundingRect.left - containerBoundingRect.left}px) scaleX(${tabBoundingRect.width})`;
```

This relies on a bit of extra CSS I’ll post in my next comment, but basically we set `width` to a static value of `1px`, and then control the actual width using a scale — still the same maths as before but more performant! Similarly `translateX` replaces our `left` property.

Reply:
Good catch! I completely forgot about trqnsforms here!